includes/class-wc-gateway-stripe.php

```diff
@@ -933,4 +962,40 @@ public function display_order_payout( $order_id ) {
 
 		<?php
 	}
+
+	public function prepare_source_from_intent( $user_id, $force_save_source = false ) {
+		// ToDo: This should either use or mimic WC_Stripe_Payment_Gateway::prepare_source().
+
+		if ( ! isset( $_POST['stripe_intent'] ) || empty( $_POST['stripe_intent'] ) ) { // wpcs: csrf ok.
+			throw new WC_Stripe_Exception( 'missing_intent_id', 'Missing intent ID' );
+		}
+
+		$intent_id     = wc_clean( $_POST['stripe_intent'] );     // wpcs: csrf ok.
+		$intent_object = $this->get_intent_object( $intent_id );
+
+		if ( 'requires_capture' !== $intent_object->status ) {
+			throw new WC_Stripe_Exception( print_r( $intent_object, true ), 'The payment intent is not expecting a capture.' );
+		}
+
+		$response = (object) array(
+			// 'token_id'      => $wc_token_id, // This was in place, investigate where it is used...
+			'customer'      => new WC_Stripe_Customer( $user_id ),
+			'source'        => $intent_object->source,
+			'source_object' => self::get_source_object( $intent_object->source ),
```

Comment:
You can save this extra request to Stripe by passing `'expand' => array( 'source' )` to the "retrieve Payment Intent" API call. See [expanding objects](https://stripe.com/docs/api/expanding_objects).

Reply:
I saw some differences in the response within `stripe.js` before and subconsciously avoided using the expanded response here.

Now that you mentioned it, I checked and I see that there is no reason to avoid `expand`, so I modified the code to use it.